### ISSUE TITLE
Ignore rules are not respected if a file in an ignored path is force added

### ISSUE THREAD

**Issue by mcdado** [libgit2/libgit2#3535]
While using [GitUp](http://github.com/git-up/GitUp) I noticed some strange behaviors that I reported to @swisspol, the author of GitUp.

I reported the problem on the [GitUp forum](http://forums.gitup.co/t/nested-gitignore-with-pattern-based-exceptions-not-working-correctly/334/5?u=mcdado) but it seems to be a libgit2 bug, which GitUp uses to interface with the repository and .gitignore files.

GitUp uses a slightly forked version of libgit2 which is based on e0ab1ca0288db57fabbd63dfb6a0cd27f3acf8b2 as of GitUp b1009 (which still shows the problem).

I created a [demo repo](/mcdado/nestedignores-demo) on GitHub to showcase the problem.

I will use GitUp as a reference, but its behavior should depend on libgit2 as far as I know, so please excuse me if I use its own behavior as the culprit.

Here are the steps that I followed (details below):
![screen shot 2015-12-06 at 00 19 17](https://cloud.githubusercontent.com/assets/898057/11613177/f5cb923e-9c17-11e5-8984-a6e79abaaf65.png)

1. Create and commit a `.htaccess` and `.gitignore` file in the root of the project. `.gitignore` ignores everything and then adds some exceptions, for itself and for the `.htaccess` file.
2. Create a `test/` directory with a test file in it, in my case the file is `test.php`
3. Create a `.htaccess` in the `test/` directory with some dull content in it

So far so good. Now things start to get messy.

4. Force add `test/test.php` even if it's in a ignored path and commit

After adding a file in the `test/` directory that is ignored, now the base `.gitignore` rule is applied to this directory too, and GitUp shows the .htaccess file.
![screen shot 2015-12-06 at 13 08 07](https://cloud.githubusercontent.com/assets/898057/11613250/a847b31e-9c1a-11e5-92a8-253582d897e2.png)

5. Try removing the `test.php` file: `git rm test/test.php`, and the `.htaccess` file disappears.
![screen shot 2015-12-06 at 13 08 41](https://cloud.githubusercontent.com/assets/898057/11613252/af834404-9c1a-11e5-94f6-85bd38f49dc5.png)

6. Try force-adding another file, for example a dummy `test/robots.txt` file, and even before committing, the `.htaccess` file appears
![screen shot 2015-12-06 at 13 08 23](https://cloud.githubusercontent.com/assets/898057/11613256/c69965a6-9c1a-11e5-8ac2-4178594c8b38.png)

As you can see in the first screenshot, git correctly ignores the files in the ignored directory (you can see a `git st` which is alias for `git status`. I originally experience this bug with a `node_modules` directory not being ignored, which is a problem because there are a ton of files in the working directory that should be ignored and create a lot of noise.

**Comment by swisspol** [comment]
I don't understand your repro steps and how they relate to the example repo you have put together. Can you indicate them in a concise way? Thanks!

For instance:
1. Clone https://github.com/mcdado/nestedignores-demo.git
2. Create file `test/foo.bar`
3. Open repo in GitUp, go to commit view: `test/foo.bar` is not ignored while it should be

**Comment by mcdado** [comment]
Yes, i'll give it another go. I used `.htaccess` as a name because that's actually what caused me problems in my setup, could be anything though.

1. `git clone git@github.com:mcdado/nestedignores-demo`
2. `cd nestedignores-demo`
3. create file `.htaccess` in test **(use `.htaccess` as name because it is ignored in `.gitignore`)**
4. open GitUp, go to Commit view. *`test/.htaccess` is not ignored while it should be.*
5. remove `test/test.php` from the repo: `git rm test/test.php`. *Now test/.htaccess is ignored as it should be.*

I hope this is exhaustive. 😉

**Comment by swisspol** [comment]
So in summary:
1. `git clone git@github.com:mcdado/nestedignores-demo`
2. `cd nestedignores-demo`
3. `touch test/.htaccess`
4. `git status` reports no modified file while libgit2 does

The `.gitignore` file contains:
```
/*

!.gitignore

!.htaccess
```

**Comment by carlosmn** [comment]
This looks like an odd edge-case git. You should not be able to unignore `test/.htaccess` since you never added it to the ignore list. But does glob-adding a directory count? Ignoring `test/` doesn't let you unignore `test/.htaccess`, so it seems like this shouldn't either.

Does git test this particular scenario?

**Comment by swisspol** [comment]
Another related but different repro case from @mcdado :

- `git clone git@github.com:mcdado/nestedignores-demo` and `cd nestedignores-demo`
- Create a new folder: `foo/`, inside of that create two folders: `one/` and `two/`
- In each of them, create a file `package.txt` and a folder called `rubbish/`, in each `rubbish/` folder create a file called `garbage.txt`.
- Add the following two lines to the end of `.gitignore`:
```
    rubbish/
    !/foo/
```
In this situation `git` ignores correctly the `rubbish/` folders (and their content), but GitUp shows them in the working directory. Trying to run `git add foo` would only add `one/package.txt` and `two/package.txt` (do not add yet, unstage the folder for now).
- Add `/foo/two` at the end of `.gitignore`
The folder `two/` is correctly ignored. `git add foo` will only add `one/package.txt` and not `one/rubbish`, but GitUp stil shows `one/rubbish/garbage.txt`
- In file `.gitignore` move the line `rubbish/` at the end
This does "solve" the problem, which leads to believe that the rules are not applied logically.

**Comment by mcdado** [comment]
@carlosmn FWIW, adding `test/` to the ignore rules has different results whether is placed above or below the .htaccess rule: if `test/` comes before `!.htaccess`, then libgit will report .htaccess while git doesn't. Placing it afterwards it completely ignore that folder.

Not knowing anything about git internals or libgit internals for that matter, it seems to me that it is a question of rule priority and overriding. I'm available for more tests.

**Comment by chrisant996** [comment]
Another example of this:

.gitignore:
```
build*.log
/*
!/*
```

In libgit2, the `!/*` cancels the `build*.log`.
But in git, the `!/*` doesn't cancel the `build*.log`.

**Comment by mcdado** [comment]
@carlosmn Hey there, it's been one year since I opened this, I was wondering what's the situation. It has been tagged "git change", what does it mean? :)

**Comment by romkatv** [comment]
FYI: https://github.com/romkatv/libgit2/commit/a546b232b23814de9c561fc750791363a5ed347e contains two self-contained test cases for this bug and a fix.

The short story is that libgit2 sometimes drops negative patterns from `.gitignore` files on the floor during parsing of said files. The code appears to be written with the assumption that the dropped negative patterns are redundant and thus can be dropped to improve performance without any user-visible side effects. When these assumptions are incorrect, libgit2 ends up ignoring some files that shouldn't be ignored.

**Test Case 1**

```zsh
rm -rf /tmp/repo
mkdir /tmp/repo
cd /tmp/repo
git init

mkdir foo
touch foo/bar

cat >.gitignore <<END
*
!foo
END

cat >foo/.gitignore <<END
!bar
END
```

(Assume that https://github.com/libgit2/libgit2/pull/5210 is merged.)

Here, when `foo/.gitignore` is parsed by libgit2, pattern `!bar` gets dropped as if it didn't exist. As a result, libgit2 will think that `foo/bar` is ignored while it shouldn't be.

**Test Case 2**

```zsh
rm -rf /tmp/repo
mkdir /tmp/repo
cd /tmp/repo
git init

mkdir foo
touch foo/bar

cat >.gitignore <<END
*
!foo
!foo/bar
END
```

In this case `!foo/bar` gets dropped, and `foo/bar` once again ends up begin ignored by libgit2 when it shouldn't be.

The fix in https://github.com/romkatv/libgit2/commit/a546b232b23814de9c561fc750791363a5ed347e stops dropping negative patterns altogether. It's possible to keep the optimization (that is, to continue dropping negative patterns that have no effect), but this requires consulting all parent `.gitignore` files when parsing a nested `.gitignore`. This isn't too bad, but we'll also need to invalidate all child `.gitignore` files when top-level `.gitignore` changes. In the end I opted for the simple fix of yanking some code out.

**Comment by buddyspike** [comment]
@romkatv, thanks! I also wondered how important it is to optimize these rules away. I tried to fix this by taking the rules in other `.gitignore` files into consideration. However, it was not that trivial because one code path discovers `.gitignore` in top-down order and the other one does it in bottom-up.
In #5210 discussion, @ethomson suggested that we should standardize it to always do `top-down`. I couldn't find some time to finish it.

**Comment by pks-t** [comment]
@romkatv: I've long wanted to completely drop this optimization of dropping the negative patterns. It's burdensome, hard to understand and often leads to wrong results. It seems like your fix cherry-picked as a standalone breaks two tests ignore::status::negative_ignores_inside_ignores and ignore::status::negative_directory_ignores. Have you taken a look at these?

**Comment by pks-t** [comment]
I think the issues stem from another discrepancy between git and libgit2: when collecting gitignores for a certain path, git will traverse top-down and abort as soon as one directory in that chain is aborted. In contrast to that, we are traversing bottom-up and will thus include all gitignores. I'll have a look

**Comment by romkatv** [comment]
> (Assume that https://github.com/libgit2/libgit2/pull/5210 is merged.)

This part of my comment is relevant.

**Comment by romkatv** [comment]
> But #5210 is merged already :/

Oh, indeed. I can confirm that these two tests are also failing in my fork. Curiously, gitstatusd handles these cases correctly, which is the reason I haven't noticed any issues with https://github.com/romkatv/libgit2/commit/a546b232b23814de9c561fc750791363a5ed347e. This might be because gitstatusd uses pathspecs, or it could be something else. I'll dig into it when I find time, although it's not very high priority for me given that gitstatusd appears to work correctly.
